expan/core/categorical_binning.py

```diff
@@ -0,0 +1,36 @@
+from heapq import heapify, heappush, heappop
+
+def categorical_binning(items, n_bins):
+    """ Performs greedy (non-optimal) binning
+    :param items: list of items to bin according to their frequencies
+    :return: list of bins [(total_weight, [items])]
+    """
+
+    # count items
+    weights = {}
+    for item in items:
+        if not item in weights:
+            weights[item] = 1
+        else:
+            weights[item] += 1
+
+    # we need items sorted in decreasing order
+    pairs = sorted([(weight, [item]) for (item, weight) in weights.items()], reverse=True)
+
+    # take max(n_bins, len(pairs)) haviest pairs as initial bins
+    bins = [(weight, labels) for (_, (weight, labels)) in zip(range(n_bins), pairs)]
```

Comment:
is it the same as pairs[:n_bins] ?

Reply:
not quite.

`pairs` might happen to be have fewer than `n_bin` elements (`zip` would terminate at the shortest of two)